upstairs/src/client.rs

```diff
@@ -0,0 +1,3127 @@
+    pub(crate) fn should_do_more_work(&self) -> bool {
+        self.client_task.is_some()
+            && !self.new_jobs.is_empty()
+            && matches!(self.state, DsState::Active | DsState::LiveRepair)
+            && (crate::MAX_ACTIVE_COUNT
+                - self.io_state_count.in_progress as usize)
+                > 0
+    }
+
+    /// Choose which `ClientAction` to apply
+    ///
+    /// This function is called from within a top-level `select!`, so not only
+    /// must the select expressions be cancel safe, but the **bodies** must also
+    /// be cancel-safe.  This is why we simply return a single value in the body
+    /// of each statement.
+    pub(crate) async fn select(&mut self) -> ClientAction {
+        tokio::select! {
+            d = async {
+                if let Some(c) = &mut self.client_task {
+                    tokio::select! {
+                        m = c.client_response_rx.recv() => {
+                            match m {
+                                Some(ClientResponse::Connected) =>
+                                    ClientAction::Connected,
+                                Some(ClientResponse::Message(m)) =>
+                                    ClientAction::Response(m),
+                                Some(ClientResponse::Done(r)) =>
+                                    ClientAction::TaskStopped(r),
+                                None =>
+                                    ClientAction::ChannelClosed,
+                            }
+                        }
+                    }
+                } else {
+                    futures::future::pending().await
+                }
+            } => {
+                d
+            }
+            _ = sleep_until(self.ping_interval), if self.client_task.is_some()
+            => {
+                ClientAction::Ping
+            }
+            _ = sleep_until(self.timeout_deadline) => {
+                ClientAction::Timeout
+            }
+        }
+    }
+
+    /// If the client task is running, send a `Message::HereIAm`
+    ///
+    /// If the client task is **not** running, log a warning to that effect.
+    pub(crate) async fn send_here_i_am(&mut self) {
+        if let Some(task) = &self.client_task {
+            if let Err(e) = task
+                .client_request_tx
+                .send(Message::HereIAm {
+                    version: CRUCIBLE_MESSAGE_VERSION,
+                    upstairs_id: self.cfg.upstairs_id,
+                    session_id: self.cfg.session_id,
+                    gen: self.cfg.gen,
+                    read_only: self.cfg.read_only,
+                    encrypted: self.cfg.encrypted(),
+                    alternate_versions: vec![],
+                })
+                .await
+            {
+                warn!(self.log, "failed to send HereIAm: {e}");
+                // TODO we should probably fail more loudly here?
+            }
+        } else {
+            panic!(
```

Comment:
It can't be stopped partway through `DownstairsClient::send_here_i_am`, since we're holding `&mut self` (and there is no interior mutability at play).

Reply:
This is a trickier question than I expected, but I think that we are in the clear:
- The first thing the IO task does (once it gets into `cmd_loop`) is to send `ClientResponse::Connected`
- Anything that would stop the task has to come through that same queue (`ClientResponse::Done`)
- Therefore, we are guaranteed that the task is still running when we call `send_here_i_am`